Add `<overflow-menu>`, rename webcomponents

silverwind opened the pull request:
1. Add `<overflow-menu>` web component
2. Rename `<gitea-origin-url>` to `<origin-url>` and make filenames match.

<img width="439" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/2fbe4ca4-110b-4ad2-8e17-c1e116ccbd74">

<img width="444" alt="Screenshot 2024-03-02 at 21 36 52" src="https://github.com/go-gitea/gitea/assets/115237/aa8f786e-dc8c-4030-b12d-7cfb74bdfd6e">

<img width="537" alt="Screenshot 2024-03-03 at 03 05 06" src="https://github.com/go-gitea/gitea/assets/115237/fddd50aa-adf1-4b4b-bd7f-caf30c7b2245">

![image](https://github.com/go-gitea/gitea/assets/115237/0f43770c-834c-4a05-8e3d-d30eb8653786)

![image](https://github.com/go-gitea/gitea/assets/115237/4b4c6bd7-843f-4f49-808f-6b3aed5e9f9a)

TODO:

- [x] Check if removal of `requestAnimationFrame` is possible to avoid flash of content. Likely needs a `MutationObserver`.
- [x] Hide tippy when button is removed from DOM.
- [x] ~~Implement right-aligned items (https://github.com/go-gitea/gitea/pull/28976)~~. Not going to do it.
- [x] Clean up CSS so base element has no background and add background via tailwind instead.
- [x] Use it for org and user page.

silverwind commented:
Works somewhat, still has some quirks where it does not move items into menu. Also there is a unavoidable flash on page load, likely could be avoided by moving into web component, but I'm not really familiar with those.

silverwind commented:
Many quirks fixed, few more to go. Already works pretty well now.

silverwind commented:
It's now a web component and I also overhauled the naming and file structure of our web components:

- Filename now matches component name.
- Removed unnecssary `gitea-` prefix.
- Renamed `webcomponents.js` to `index.js`.
- Moved all polyfills to `polyfills.js`.

Two TODO items left mentioned above until this is ready.

@wxiaoguang hope you like those renames 😆.

denyskon commented:
Thanks for your great work! Could you check if this would be easily addable to org/user profile menu? If not, we can leave it out for a follow-up PR :)

silverwind commented:
> Thanks for your great work! Could you check if this would be easily addable to org/user profile menu? If not, we can leave it out for a follow-up PR :)

Shouldn't be too hard I guess, the only difference there is the 2px border on bottom.

silverwind commented:
Right-aligned items will be a bit of a challenge because it needs to detect and measure those items separately. I'm not sure if worth the trouble and just left-align everything, just like GitHub has, including on the org page for "Settings".

denyskon commented:
I would simply not do right-aligned items then ..

wxiaoguang commented:
> @wxiaoguang hope you like those renames 😆.

TBH, I don't like. Sooner or later they will conflict with something. I am not a fan of creating conflicts.

silverwind commented:
> > @wxiaoguang hope you like those renames 😆.
> 
> TBH, I don't like. Sooner or later they will conflict with something. I am not a fan of creating conflicts.

We are in full control which external custom elements we install, I see no issue. Also conflicts are not possible because one can not register the same custom element twice without it [throwing an error](https://jsfiddle.net/silverwind/r2nxvmqu/).
```

silverwind commented on web_src/js/modules/tippy.js:
```diff
@@ -40,13 +40,6 @@ export function createTippy(target, opts = {}) {
-  if (other.content instanceof Element) {
-    other.content.classList.remove('tippy-target');
-  }
-
```
FYI for reviewers: This mechanism was moved to CSS, `.tippy-target` will hide the element like before, and `.tippy-content .tippy-target` will show it, so this class removal isn't necessary anymore.

silverwind commented:
Org page working:

<img width="445" alt="Screenshot 2024-03-03 at 02 40 10" src="https://github.com/go-gitea/gitea/assets/115237/dfa52ef9-026c-4d21-b418-59cf52ce308f">

silverwind commented:
User page working:

<img width="569" alt="Screenshot 2024-03-03 at 02 43 21" src="https://github.com/go-gitea/gitea/assets/115237/60b4e3e9-afff-4cfc-a15f-b0500cc4873d">

silverwind commented:
FYI this is with 36px threshold where it will hide a element:

<img width="115" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/32662828-d603-422f-94c9-8dd13138dbc1">

To hide at the outer edge of the padding, something like 50px would be necessary but I think this is acceptable that the button bleeds slightly into the element's padding.

silverwind commented:
Ready now. Got rid of `requestAnimationFrame`, so there are no more loading flashes. The component waits for a mutation event for the mandatory `.overflow-menu-items` and then does its thing.

@lunny take note we decided against right-aligned items and the component does not currently support those, so "Settings" is left-aligned everywhere now.

silverwind commented:
Side note: the element could likely be used for `#navbar` as well, replacing a lot of hacks around the fomantic menu and the mobile hamburger button, but I'll leave that to a future PR.

lunny commented:
> Ready now. Got rid of `requestAnimationFrame`, so there are no more loading flashes. The component waits for a mutation event for the mandatory `.overflow-menu-items` and then does its thing.
> 
> @lunny take note we decided against right-aligned items and the component does not currently support those, so "Settings" is left-aligned everywhere now.

I'm OK if all the places are consistent whatever left or right.

silverwind commented:
https://github.com/go-gitea/gitea/pull/29400/commits/6182d5f229f1e2142b01e51885c90a9b2c6fc2aa includes:

- Make Vue [aware of our custom elements](https://vuejs.org/api/application.html#app-config-compileroptions-iscustomelement) so it doesn't try to resolve them as vue components.
- Make initialization compatible with rendering in Vue where the child element is immediately present and no mutation events are triggered. I think this is good compat to have, even thought I did observe all browsers triggering the mutations even when rendering outside of Vue.
- Add hover color to "..." button

silverwind commented:
There appears to be some FOUC issue, visible only in Firefox, on pages that have the element. Not sure how to prevent, maybe I have to move it back regular `div` element.

silverwind commented:
Ended up splitting the webcomponents chunk into two which seems to have gotten rid of all FOUCs (with `defer`, the `<relative-time>` would FOUC). The file is a bit too big currently, will check what can be done to reduce it:

```
180K public/assets/js/webcomponents.defer.js
 34K public/assets/js/webcomponents.js
```

silverwind commented:
The FOUC in Firefox is triggered by the `getBoundingClientRect` calls during initial rendering, I think these need to ideally be eliminated or delayed on initial render.

silverwind commented:
All fixed now including the Firefox FOUC which apparently happend because DOM mutation was done too soon during rendering. I was able to remove the initial `updateItems` by just relying on the ResizeObserver, which as per its spec will trigger a callback during initial rendering.

Also, I did a number of performance improvements to the component as well, removing a number of `querySelector` and removing unnecessary `cloneNode`.

silverwind commented on .eslintrc.yaml:
```diff
@@ -828,7 +828,7 @@ rules:
   wc/max-elements-per-file: [0]
   wc/no-child-traversal-in-attributechangedcallback: [2]
-  wc/no-child-traversal-in-connectedcallback: [2]
+  wc/no-child-traversal-in-connectedcallback: [0] # neccesary for custom elements in Vue
```
```suggestion
  wc/no-child-traversal-in-connectedcallback: [0] # neccessary for custom elements in Vue
```

silverwind commented on .eslintrc.yaml:
```diff
@@ -828,7 +828,7 @@ rules:
   wc/max-elements-per-file: [0]
   wc/no-child-traversal-in-attributechangedcallback: [2]
-  wc/no-child-traversal-in-connectedcallback: [2]
+  wc/no-child-traversal-in-connectedcallback: [0] # neccessary for custom elements in Vue
```
```suggestion
  wc/no-child-traversal-in-connectedcallback: [0] # needed for custom elements in Vue where children are available during connectedCallback
```

denyskon commented:
Could you please also add it here?
![grafik](https://github.com/go-gitea/gitea/assets/47871822/7c793d99-9ad6-4cd6-8e21-fa5bcd252880)

Works really great, thanks again for this overdue change :)

silverwind commented:
> Could you please also add it here? ![grafik](https://private-user-images.githubusercontent.com/47871822/309557995-7c793d99-9ad6-4cd6-8e21-fa5bcd252880.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk0Nzc1NzIsIm5iZiI6MTcwOTQ3NzI3MiwicGF0aCI6Ii80Nzg3MTgyMi8zMDk1NTc5OTUtN2M3OTNkOTktOWFkNi00Y2Q2LThlMjEtZmE1YmNkMjUyODgwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzAzVDE0NDc1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRkZDNlZjM5YTViYjZiODY3MWVlNjA2YTFhY2Q3YzNiMjgxMGM3Yzc4YmVhMDk4MjIyZmIyNWFjZGU2YWYyODAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.t0kgAvpNjxVTuAFRCJhvZrsaCdv9raXHflyqA7C46VU)
> 
> Works really great, thanks again for this overdue change :)

I tried there but didn't like the visual outcome so much there, but I can try again.

silverwind commented:
I will give it a quick try, If I recall correctly, the issue was that the "..." button was not aligning so well, but it can likely be fixed, or is already fixed since I changed it to flexbox some commits ago.

silverwind commented:
Now using it there too. I also moved all relevant styles to scoped CSS.

<img width="259" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/27b76533-f5cc-412f-8220-73fa0a158d46">

silverwind commented:
One issue remains: There is this browser "warning" when switching to org tab and back to repo:

<img width="1333" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/c244e763-bf81-4956-b5e1-1b66b2849321">

This is only visible in dev mode, but still needs to be fixed.

silverwind commented:
> One issue remains: There is this browser "warning" when switching to org tab and back to repo on the repo list:
> <img alt="image" width="1333" src="https://private-user-images.githubusercontent.com/115237/309561364-c244e763-bf81-4956-b5e1-1b66b2849321.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk0ODE0OTksIm5iZiI6MTcwOTQ4MTE5OSwicGF0aCI6Ii8xMTUyMzcvMzA5NTYxMzY0LWMyNDRlNzYzLWJmODEtNDk1Ni1iNWUxLTFiNjZiMjg0OTMyMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMzAzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDMwM1QxNTUzMTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mYmVkMDU0ZWVmNWU4ZDM0MzY5OTJmOTYwY2RlMDkyMTkxYWIwNmQ5NTMyMjc3ZGNjN2U2YzA2ZjE4ZDNjMjIwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.pOTClKiuIHk0BtNAw2QuTX0kgCFXpUp2aRB0tcFLC8c">
> 
> This is only visible in dev mode, but still needs to be fixed.

Fixed in https://github.com/go-gitea/gitea/pull/29400/commits/77a8e1e9c08ec66fed296aad8e1efadca08115bd.

silverwind commented:
Maybe some prefix on webcomponents will still be good to have, if only to not having to maintain [this array](https://github.com/go-gitea/gitea/pull/29400/files#diff-1fb26bc12ac780c7ad7325730ed09fc4c2c3d757c276c3dacc44bfe20faf166fR124). Maybe `wc-`.

denyskon commented:
`wc` looks good I think

silverwind commented:
Renamed to `wc-` prefix and added it to docs.

silverwind commented:
Few more things done:

- Updated webcomponents README
- Added all used webcomponents including external ones to webpack config for Vue
- Imported the Icon via `import` instead of inlining the source

wxiaoguang commented:
Could there be 2 separate PRs, one for component naming, one for the "overflow-menu"? TBH, mixing them together seems difficult to review .... (ps: haven't got enough time to review the "overflow-menu" related code today, maybe I would be quite busy till this weekend).

silverwind commented:
I thought about splitting into PRs, but decided against it because it would create dependencies between the PRs for consistency, so I prefer it to do it in one PR. Sorry for the inconvenience.

`<wc-overflow-menu>` code should be pretty straightfoward to review. the mechanism is to always append all items to the menu, measure which ones are outside the element, and move those to the tippy. It's currently throttled to run every 100ms at most, but it seems performant enough to even run unthrottled.

There is one tiny issue remaining: When resizing the window continuously, there is one point where the last element may flash in and out a bit, I will try to fix that, but it's definitely not a blocker imho because as soon as you let go of the resize handle, it will stabilize.

silverwind commented on web_src/js/webcomponents/wc-overflow-menu.js:
```diff
@@ -0,0 +1,109 @@
+    const menuRight = this.offsetLeft + this.offsetWidth;
+    for (const item of this.menuItemsEl.querySelectorAll('.item')) {
+      const itemRight = item.offsetLeft + item.offsetWidth;
+      if (menuRight - itemRight < 38) { // slightly less than width of .wc-overflow-menu-button
```
```suggestion
      if (menuRight - itemRight < 38) { // roughly the width of .wc-overflow-menu-button
```

silverwind commented:
If you really prefer, I can split. There is one more change I want to do, which is the removal of the `@webcomponents/custom-elements` polyfill as PaleMoon now supports custom elements since v32.1.0 (2023-03-21).

wxiaoguang commented on web_src/js/webcomponents/wc-overflow-menu.js:
```diff
@@ -0,0 +1,109 @@
+            if (node.classList.contains('wc-overflow-menu-items')) {
+              this.menuItemsEl = node;
+              this.mutationObserver?.disconnect();
+              this.init();
```
Duplicate init? It seems that `new ResizeObserver` would be called many times if there are many mutations

wxiaoguang commented on web_src/css/modules/tippy.css:
```diff
@@ -56,6 +61,15 @@
   fill: var(--color-menu);
 }
 
+.tippy-box[data-theme="menu"] .tippy-content > .ui.menu {
+  width: auto; /* unset fomantic's default 15rem */
+  border: none; /* remove `.ui.menu` border as it is provided by tippy */
+}
+
+.tippy-box[data-theme="menu"] .tippy-content .ui.vertical.menu .item::before {
+  content: none; /* remove lines between items */
+}
```
Is it possible to get rid of the Fomantic defined classes? I do not think it is maintainable to mix different frameworks together.

silverwind replied:
Likely, but then I need to come up with a few new rules to style menus in tippy. But good idea, I will check. This will affect also the "line button" in code view.

silverwind replied:
I think I may re-purpose the `.item`class, as I assume fomantic would not touch it if it's standalone with no `.menu` parent.

silverwind replied:
Done, removed the dependency from both menus:

<img width="201" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/0c946955-11d9-45e2-aa67-c956c8efed03">

<img width="367" alt="image" src="https://github.com/go-gitea/gitea/assets/115237/7ae81513-a465-4e16-8c44-8dd07cfcf98d">

wxiaoguang commented:
And one more thing, is it screen-reader friendly (especially for the popup, it is always a complex topic in a11y field)? There have been a lot of a11y problems in Gitea, if there is a chance, maybe we could slightly improve them, at least, avoid make the a11y worse.

silverwind commented:
Maybe I will actually revert the `wc-` prefix because the webpack config isn't much simpler without it, external components will never follow such naming schemes and GitHub's web components are all without any prefix as well.

silverwind commented:
> And one more thing, is it screen-reader friendly (especially for the popup, it is always a complex topic in a11y field)? There have been a lot of a11y problems in Gitea, if there is a chance, maybe we could slightly improve them, at least, do the best to avoid making the a11y worse.

Added `aria-haspopup` for all `click` trigger tippies, added `aria-label` to the button and `aria-role` to menu items while they are in the tippy. `aria-expanded` is already handled by tippy.

wxiaoguang commented:
1. `  if (trigger === 'click')` then set `haspopup=menu` seems quite strange to me.
    * Why not use "role" ? The components should know what role it should be, but not guessing by a event trigger
    * I am not a fan of keeping "guessing" things.
2. Does the aria really work? According to my quick test, at least it doesn't work on desktop (I haven't got time to test it on mobile).
    * I have written some documents in `aria.md`
    * By using VoiceOver, if the "aria menu" is correctly constructed, it should be able to navigate items by Option+Up/Down. 

ps: I might be away from keyboard in the following days, maybe no enough time to do deep tests until some days later.

silverwind commented:
> Why not use "role" ? The components should know what role it should be, but not guessing by a event trigger

Yes, role might work. I misinterpreted the [docs](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-haspopup) to think it is meant for clicks, but should work for non-clicks too.

> Does the aria really work? According to my quick test, at least it doesn't work on desktop (I haven't got time to test it on mobile). 

I did not actually test it, but I can try VoiceOver I guess.

> I am not sure whether additional_options = Additional options is the proper sentence. IMO in many cases, there are more "actions" or "items" or "links", but not "options".

Github uses "Additional navigation options", but because our case, the menu is not used for navigation only. Not sure how to name it. "More items" may be good and I think it already has translations as well.

denyskon reviewed (approved):
![image](https://github.com/go-gitea/gitea/assets/47871822/32dd8c71-d365-4acb-9dff-74004c273185)
Could this arrow be somehow highlighted too on hover?

silverwind commented:
> ![image](https://private-user-images.githubusercontent.com/47871822/311427850-32dd8c71-d365-4acb-9dff-74004c273185.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk5ODc1OTcsIm5iZiI6MTcwOTk4NzI5NywicGF0aCI6Ii80Nzg3MTgyMi8zMTE0Mjc4NTAtMzJkZDhjNzEtZDM2NS00YWNiLTlkZmYtNzQwMDRjMjczMTg1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzA5VDEyMjgxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA0ODE4MWEyNWY3M2I3MWE2NGRiNTljZWYxNzAyMDY5MDY2Y2RkNWE5ODAxMjgzMjg3ODUzOTllZTNiNTYwODEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PqluLsX4VvFEtDtgkXX6lwxxA_jBvvJrrkU-4NmzfSQ) Could this arrow be somehow highlighted too on hover?

Yes, I had that somewhat working already via a somewhat complex CSS selector but the main issue is to to obtain the correct color, because `--color-hover` is partially transparent. I tried using `color-mix`, but not matter which color space I chose, it just wouldn't exactly match.

denyskon commented:
That would work too I guess :)

silverwind commented:
Padding added: 

<img width="146" alt="Screenshot 2024-03-12 at 22 52 21" src="https://github.com/go-gitea/gitea/assets/115237/3b23c5e8-063d-4ff2-b6fb-b3ddca90c230">

silverwind commented:
Small unrelated tweak: Added box-shadow to this:

<img width="353" alt="Screenshot 2024-03-12 at 23 16 35" src="https://github.com/go-gitea/gitea/assets/115237/dc7b5f88-6ac8-479c-a65f-7253d9914bf9">

silverwind commented:
Keyboard navigation is missing from these tippy menus, but honestly I don't know how to do those properly, so I would defer that for later.

wxiaoguang commented:
I managed to add the initial aria support in 9c2fdb97353f3307208ebe98e6a76ce25acc6a0f , not ideal enough but indeed it could be fully operated by keyboard & screenreader (use Tab/Space, no arrow key yet).

I would leave the final reviews and decisions to TOC.

silverwind commented:
Will merge and include the rename for `<absolute-time>`, then it should definitely be ready.

silverwind commented:
Keyboard interaction now fully works: Arrow keys, Tab, Shift+Tab, Enter, Spacebar.

silverwind commented:
Ready again for review. Hope the renames don't pollute the diff too much 😉.

wxiaoguang commented on web_src/js/webcomponents/overflow-menu.js:
```diff
@@ -22,18 +23,34 @@ window.customElements.define('overflow-menu', class extends HTMLElement {
-          if (document.activeElement.matches('[role="menuitem"]')) {
-            e.preventDefault();
+        } else if (e.key === ' ' || e.code === 'Enter') {
+          e.preventDefault();
+          e.stopPropagation();
```
It shouldn't "prevent" too early, for example, maybe there will be "input" in the popup

silverwind replied:
Ah, I was replicating from [here](https://github.com/tailwindlabs/headlessui/blob/626a253dcf09589e2a2fbbfd0c279120153394cc/packages/%40headlessui-vue/src/components/menu/menu.ts#L257), but I see it's likely better to do conditionally.

silverwind replied:
fixed with https://github.com/go-gitea/gitea/pull/29400/commits/97ed9ed33f871d36c5885f50ccc868d5fe6383bf

wxiaoguang commented on web_src/js/webcomponents/overflow-menu.js:
```diff
@@ -22,18 +23,34 @@ window.customElements.define('overflow-menu', class extends HTMLElement {
+            document.activeElement.nextElementSibling?.focus();
+          }
+        } else if (e.key === 'ArrowUp') {
+          e.preventDefault();
+          e.stopPropagation();
```
The same as above, it should only "prevent" if and only if the event is really handled by us.

silverwind replied:
fixed with https://github.com/go-gitea/gitea/pull/29400/commits/97ed9ed33f871d36c5885f50ccc868d5fe6383bf
